File: 0290-word-pattern/0290-word-pattern.rb
# @param {String} pattern
# @param {String} s
# @return {Boolean}
def word_pattern(pattern, s)
  words = s.split
  return false if pattern.length != words.length
  fw, bw = Hash.new, Hash.new
  (0...pattern.length).each do |i|
    if !fw.include?(pattern[i]) && !bw.include?(words[i])
      fw[pattern[i]] = words[i]
      bw[words[i]] = pattern[i]
    elsif fw.include?(pattern[i]) && !bw.include?(words[i])
      return false
    elsif !fw.include?(pattern[i]) && bw.include?(words[i])
      return false
    elsif fw[pattern[i]] != words[i] || bw[words[i]] != pattern[i]
      return false
    end
  end
  true
end

File: 0290-word-pattern/0290-word-pattern.ts
function wordPattern(pattern: string, s: string): boolean {
  const words: string[] = s.split(' ')
  if (pattern.length !== words.length) return false
  const fw = {}
  const bw = {}
  for (let i = 0; i < pattern.length; i++) {
    if (words[i] === "constructor") words[i] = "constructor-"
    if (!(pattern[i] in fw) && !(words[i] in bw)) {
      fw[pattern[i]] = words[i]
      bw[words[i]] = pattern[i]
    } else if (pattern[i] in fw && !(words[i] in bw)) {
      return false
    } else if (!(pattern[i] in fw) && words[i] in bw) {
      return false
    } else if (fw[pattern[i]] !== words[i] || bw[words[i]] !== pattern[i]) {
      return false
    }
  }
  return true
}